### PR TITLE
VSCode Extension for Ballerina Integration

### PR DESCRIPTION
**Pull request by IrushiL**
- Capability to view integration templates available. 
- Capability to create a new project using an integration template available. 
- Provide code snippet support.

Related Issue: https://github.com/wso2/ballerina-integrator/issues/39
Resolves: https://github.com/wso2/ballerina-integrator/issues/97

### CODE REVIEW

File: vscode/.gitignore [hunk 1]
```diff
@@ -0,0 +1,4 @@
+out
```
Comment (pramodya1994) [inline vscode/.gitignore]: Shall we add all git ignore codes to root .gitignore file?

Comment (IrushiL) [reply vscode/.gitignore]: Done

File: vscode/src/createProject.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+export async function createTemplateProject(currentPanel: vscode.WebviewPanel, context: vscode.ExtensionContext) {
+    let templateSelected = undefined;
+    let projectTemplates = new ProjectTemplates(context, workspace.getConfiguration('projectTemplates'));
+    // get workspace folder
+    let workspaceSelected = await projectTemplates.selectWorkspace();
+    window.showInformationMessage("Workspace Selected: " + workspaceSelected);
+    // generate the home page to display templates
+    currentPanel.webview.html = getHomeView();
+    currentPanel.webview.onDidReceiveMessage(
+        messageA => {
```
Comment (pramodya1994) [inline vscode/src/createProject.ts]: messageA?

File: vscode/src/createProject.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            if (!workspaceSelected) {
+                window.showErrorMessage("No Workspace Selected!");
+                return;
+            }
+            templateSelected = messageA.command;
+            // generate the placeholder form for a specific template
+            currentPanel.webview.html = getFormView(templateSelected);
+            currentPanel.webview.onDidReceiveMessage(
+                async messageB => {
```
Comment (pramodya1994) [inline vscode/src/createProject.ts]: messageB?

File: vscode/src/extension.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import vscode, { commands, window, ViewColumn } from 'vscode';
+import { createTemplateProject } from './createProject';
+
+// this method is called when your extension is activated
+// your extension is activated the very first time the command is executed
+export function activate(context: vscode.ExtensionContext) {
+
+	// Use the console to output diagnostic information (console.log) and errors (console.error)
+	// This line of code will only be executed once when your extension is activated
+	console.log('Congratulations, your extension "ballerina-integrator" is now active!');
```
Comment (pramodya1994) [inline vscode/src/extension.ts]:
```suggestion
	console.log('Your extension "ballerina-integrator" is now active!');
```

File: vscode/src/createProject.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                    if (messageB.command == "back") {
```
Comment (pramodya1994) [inline vscode/src/createProject.ts]:
```suggestion
                    if (messageB.command === "back") {
```

File: vscode/src/createProject.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+                        currentPanel.dispose();
+                        vscode.commands.executeCommand("ballerinaIntegrator.projectTemplates");
+                        return;
+                    } else {
+                        projectTemplates.updateConfiguration(workspace.getConfiguration('projectTemplates'));
+                        let templateObject = data.find(x => x.id == messageA.command);
```
Comment (pramodya1994) [inline vscode/src/createProject.ts]:
```suggestion
                        let templateObject = data.find(x => x.id === messageA.command);
```

File: vscode/src/createProject.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+                        projectTemplates.createFromTemplate(workspaceSelected, messageA.command, placeholders).then(
+                            (template: string | undefined) => {
+                                if (template) {
+                                    window.showInformationMessage("New template project created for '" +
+                                        templateObject.name + "'!");
+
+                                }
+                            },
+                            (reason: any) => {
+                                if (reason == "false") {
```
Comment (pramodya1994) [inline vscode/src/createProject.ts]:
```suggestion
                                if (reason === "false") {
```

File: vscode/src/createProject.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+                async messageB => {
+                    if (messageB.command == "back") {
+                        currentPanel.dispose();
+                        vscode.commands.executeCommand("ballerinaIntegrator.projectTemplates");
+                        return;
+                    } else {
+                        projectTemplates.updateConfiguration(workspace.getConfiguration('projectTemplates'));
+                        let templateObject = data.find(x => x.id == messageA.command);
+                        let templatePlaceholders = templateObject.placeholders;
+                        var placeholderMap = new Map();
```
Comment (pramodya1994) [inline vscode/src/createProject.ts]: var used instead of let

File: vscode/src/formView.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+import { begin, end } from './html';
+import { formStyles } from './styles';
+import data from './templateDetails.json';
+
+export function getFormView(templateSelected: string): string {
+
+    let templateObject = data.find(x => x.id == templateSelected);
```
Comment (pramodya1994) [inline vscode/src/formView.ts]:
```suggestion
    let templateObject = data.find(x => x.id === templateSelected);
```

File: vscode/src/formView.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    let templateId = templateSelected;
```
Comment (pramodya1994) [inline vscode/src/formView.ts]: local variable templateId never used.

File: vscode/src/formView.ts [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    let elements: string = "";
+    templatePlaceholders.forEach(element => {
+        elements += `
+        ${element.id}: elements[${i}].value,`;
+        i++;
+    });
+
+    let scriptHandling = `
+            <script>
+                function myFunction(template) {
```
Comment (pramodya1994) [inline vscode/src/formView.ts]: Use an appropriate name for myFunction.

Comment (pramodya1994) [reply vscode/src/formView.ts]: Unused parameter "template"

File: vscode/src/formView.ts [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+                    var elements = document.getElementById("myForm").elements;
```
Comment (pramodya1994) [inline vscode/src/formView.ts]: Unused local variable "elements"

File: vscode/src/formView.ts [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+                    })
+                }
+            </script>`;
+
+    let htmlCode = begin + formStyles + scriptHandling + beginForm;
+    let allFormElements: string = "";
+    templatePlaceholders.forEach(element => {
+        let formElement: string = "";
+        let values = element.value;
+        if (element.type != "select") {
```
Comment (pramodya1994) [inline vscode/src/formView.ts]:
```suggestion
        if (element.type !== "select") {
```

File: vscode/src/homeView.ts [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+        let card: string = `
+                <div class="column">
+                    <a href="#" style="none" onclick="pickTemplate('${templateId}')">
+                        <div class="card">
+                            <h2>` + templateName + `</h2><br/>
+                            <p>` + templateDescription + `</p>
+                        </div>
+                    </a>  
+                </div>`;
+        if (numberOfColumns % 3 == 0) {
```
Comment (pramodya1994) [inline vscode/src/homeView.ts]:
```suggestion
        if (numberOfColumns % 3 === 0) {
```

File: vscode/src/homeView.ts [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+            htmlCode = htmlCode + rowHandleStart + card;
+        }
+        else if (numberOfColumns % 3 == 2) {
```
Comment (pramodya1994) [inline vscode/src/homeView.ts]:
```suggestion
        else if (numberOfColumns % 3 === 2) {
```

File: vscode/src/projectTemplates.ts [hunk 16]
```diff
@@ -0,0 +1,329 @@
...
+    public updateConfiguration(config: WorkspaceConfiguration) {
+        this.config = config;
+    }
+
+    /**
+     * Selects a workspace folder. For single root workspaces the root directory 
+     * will be selected, or for multi-root a chooser will be presented to select a workspace.
+     * @returns The workspace selected.
+     */
+    public async selectWorkspace(): Promise<string> {
```
Comment (pramodya1994) [inline vscode/src/projectTemplates.ts]: Can't we make this method static?

File: vscode/src/projectTemplates.ts [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+            }
+        }
+        return workspaceSelected;
+    }
+
+    /**
+     * Returns a list of available project templates by reading the templates directory.
+     * @returns List of templates found.
+     */
+    public async getTemplates(): Promise<string[]> {
```
Comment (pramodya1994) [inline vscode/src/projectTemplates.ts]: Unused method getTemplates(...)

File: vscode/src/projectTemplates.ts [hunk 18]
```diff
@@ -0,0 +1,329 @@
...
+     * @param template Relative path to the template folder.
+     * @param placeholderValues Placeholder key-value pairs.
+     */
+    public async createFromTemplate(workspace: string, template: string, placeholderValues: any) {
+
+        // get template folder
+        let templateRoot = await this.getTemplatesDir();
+        let templateDir = path.join(templateRoot, template);
+        if (!fs.existsSync(templateDir) || !fs.lstatSync(templateDir).isDirectory()) {
+            window.showErrorMessage("Template '" + data.find(x => x.id == template).name + "' does not exist.");
```
Comment (pramodya1994) [inline vscode/src/projectTemplates.ts]:
```suggestion
            window.showErrorMessage("Template '" + data.find(x => x.id === template).name + "' does not exist.");
```

File: vscode/src/projectTemplates.ts [hunk 19]
```diff
@@ -0,0 +1,329 @@
...
+            return undefined;
+        }
+        // update placeholder configuration
+        let usePlaceholders = this.config.get("usePlaceholders", false);
+        let placeholderRegExp = this.config.get("placeholderRegExp", "\\${(\\w+)?}");
+        // let placeholders: { [placeholder: string]: string | undefined } = this.config.get("placeholders", {});
+        let placeholders = placeholderValues;
+        // re-read configuration, merge with current list of placeholders
+        let newplaceholders: { [placeholder: string]: string } = this.config.get("placeholders", {});
```
Comment (pramodya1994) [inline vscode/src/projectTemplates.ts]: newplaceholders => Shall we use camel case notation?

File: vscode/src/utils.ts [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+
+import fs = require("fs");
+import path = require("path");
+
+/**
+ * Recursively copy folder from src to dest.
+ * @param src Source folder.
+ * @param dest Destination folder.
+ */
+export async function copyDir(src: string, dest: string) {
```
Comment (pramodya1994) [inline vscode/src/utils.ts]: Unused function

File: vscode/src/utils.ts [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+	);
+
+	return Promise.resolve(true);
+}
+
+/**
+ * Recursively make directories.
+ * @param path Destination path.
+ */
+export function mkdirsSync(dest: string, mode: string | number | null | undefined = undefined): boolean {
```
Comment (pramodya1994) [inline vscode/src/utils.ts]: params not described in JS docs

File: vscode/src/utils.ts [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+	// check if exists
+	if (fs.existsSync(dest)) {
+		if (fs.lstatSync(dest).isDirectory()) {
```
Comment (pramodya1994) [inline vscode/src/utils.ts]: if statement can be simplified.
```
if (fs.lstatSync(dest).isDirectory()) {
	return true;
} else {
	return false;
}
```
Try to use below.
```
return fs.lstatSync(dest).isDirectory();
```

File: vscode/src/utils.ts [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+/**
+ * Converys a map into an object.
+ * @param inputMap Input map that needs to be converted.
+ */
+export function mapToObj(inputMap) {
+	let obj = {};
+	inputMap.forEach(function (value, key) {
+		obj[key] = value
```
Comment (pramodya1994) [inline vscode/src/utils.ts]:
```suggestion
		obj[key] = value;
```

File: vscode/src/createProject.ts (hunk 2 above)
Comment (IrushiL) [reply vscode/src/createProject.ts]: Changed to a meaningful name.

File: vscode/src/createProject.ts (hunk 3 above)
Comment (IrushiL) [reply vscode/src/createProject.ts]: Changed to a meaningful name.

File: vscode/src/extension.ts (hunk 4 above)
Comment (IrushiL) [reply vscode/src/extension.ts]: Done.

File: vscode/src/createProject.ts (hunk 5 above)
Comment (IrushiL) [reply vscode/src/createProject.ts]: Done.

File: vscode/src/createProject.ts (hunk 6 above)
Comment (IrushiL) [reply vscode/src/createProject.ts]: Done.

File: vscode/src/createProject.ts (hunk 7 above)
Comment (IrushiL) [reply vscode/src/createProject.ts]: Done.

File: vscode/src/createProject.ts (hunk 8 above)
Comment (IrushiL) [reply vscode/src/createProject.ts]: Done.

File: vscode/src/formView.ts (hunk 9 above)
Comment (IrushiL) [reply vscode/src/formView.ts]: Done.

File: vscode/src/formView.ts (hunk 10 above)
Comment (IrushiL) [reply vscode/src/formView.ts]: Removed.

File: vscode/src/formView.ts (hunk 11 above)
Comment (IrushiL) [reply vscode/src/formView.ts]: Done.

File: vscode/src/formView.ts (hunk 12 above)
Comment (IrushiL) [reply vscode/src/formView.ts]: This variable is used from within elements string created in the TypeScript code.

File: vscode/src/formView.ts (hunk 13 above)
Comment (IrushiL) [reply vscode/src/formView.ts]: Done.

File: vscode/src/homeView.ts (hunk 14 above)
Comment (IrushiL) [reply vscode/src/homeView.ts]: Done.

File: vscode/src/homeView.ts (hunk 15 above)
Comment (IrushiL) [reply vscode/src/homeView.ts]: Done.

File: vscode/src/projectTemplates.ts (hunk 16 above)
Comment (IrushiL) [reply vscode/src/projectTemplates.ts]: We can, changed it to a static method.

File: vscode/src/projectTemplates.ts (hunk 17 above)
Comment (IrushiL) [reply vscode/src/projectTemplates.ts]: Removed.

File: vscode/src/projectTemplates.ts (hunk 18 above)
Comment (IrushiL) [reply vscode/src/projectTemplates.ts]: Done.

File: vscode/src/projectTemplates.ts (hunk 19 above)
Comment (IrushiL) [reply vscode/src/projectTemplates.ts]: Done.

File: vscode/src/utils.ts (hunk 20 above)
Comment (IrushiL) [reply vscode/src/utils.ts]: Removed.

File: vscode/src/utils.ts (hunk 21 above)
Comment (IrushiL) [reply vscode/src/utils.ts]: Fixed.

File: vscode/src/utils.ts (hunk 22 above)
Comment (IrushiL) [reply vscode/src/utils.ts]: Done.

File: vscode/src/utils.ts (hunk 23 above)
Comment (IrushiL) [reply vscode/src/utils.ts]: Done.

**Pull request merged**